lib/remote/configobjectutility.cpp

```diff
@@ -198,13 +199,26 @@ bool ConfigObjectUtility::CreateObject(const Type::Ptr& type, const String& full
 		return false;
 	}
 
+	// AtomicFile doesn't create not yet existing directories, so we have to do it by our ourselves.
 	Utility::MkDirP(Utility::DirName(path), 0700);
 
-	std::ofstream fp(path.CStr(), std::ofstream::out | std::ostream::trunc);
+	AtomicFile fp(path, 0644);
 	fp << config;
-	fp.close();
-
-	std::unique_ptr<Expression> expr = ConfigCompiler::CompileFile(path, String(), "_api");
+	// Flush the output buffer to catch any errors ASAP and handle them accordingly!
+	// Note: AtomicFile places these configs in a temp file and will be automatically
+	//       discarded when it is not committed before going out of scope.
+	fp.flush();
+
+	// Given that the internal config::Update cluster events are using this as well to create synced runtime objects,
+	// we don't want to write the config first to the disk and then load and validate it with CompileFile.
+	// Otherwise, we are forced to remove the newly created file whenever we can't validate, commit or activate it,
+	// even if something goes terribly wrong. This approach would also have the downside that two cluster events for
+	// the same object arriving at the same time from two different endpoints would result in two threads simultaneously
+	// creating and loading the same configuration file - whereby only one of them surpasses the validation process,
+	// while the other is facing an object `re-definition` error and tries to remove that configuration file it
+	// mistakenly thinks it has created. As a consequence, an object successfully created by the former is implicitly
+	// deleted by the latter thread, causing the objects to mysteriously disappear.
+	std::unique_ptr<Expression> expr = ConfigCompiler::CompileText(path, config, String(), "_api");
```

Comment:
That comment sounds more like the reasoning for using `AtomicFile` and has little to do with `CompileText()`, so I'd move it to just before the `AtomicFile` is created.

Also, please boil down the comment to what's relevant for the new code (IMHO: this function can be called concurrently (cluster config, but probably also HTTP requests), `AtomicFile` is used so multiple threads don't affect each other and that that only the one committing the object in-memory also commits the file). How it broke before/what issue it fixes is more something for the commit message and PR description.